### PR TITLE
Feature satisfaction campaigns

### PR DESCRIPTION
**Pull request by osallou**
manager launch satisfaction surveys for a perdio with answer time limit

* email sent to user with booking or closed projects in period
* user hjas access via email link or module page to rate each resource/project with optional comment
* managers have access to each campaign with average note for type (booking and projects) and per resource (resource name or project name)

### CODE REVIEW

**Comment by osallou** [comment]
todo: add unit tests

**Comment by osallou** [comment]
@gaetanHerve ok de mon côté mais peut etre juste pour 2.7 ? (à valider / tester)

**Comment by gaetanHerve** [comment]
Je regarde ça

**Comment by osallou** [comment]
@gaetanHerve   c'est bon, j'ai apporté les corrections et fait qq tests

**Comment by osallou** [comment]
@gaetanHerve pour revue (encore) quand tu auras du temps (pas urgent pour cette release)

File: Modules/rating/Controller/RatingController.php [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+
+
+
+    public function rateAction($id_space, $id_campaign) {
+        $plan = new CorePlan($this->currentSpace['plan'], $this->currentSpace['plan_expire']);
+        if(!$plan->hasFlag(CorePlan::FLAGS_SATISFACTION)) {
+            throw new PfmAuthException('Sorry, space does not have this feature plan');
+        }
+        $userSpaceStatus = $this->getUserSpaceStatus($id_space, $_SESSION["id_user"]);
+        if($userSpaceStatus != CoreSpace::$USER) {
```
Comment (gaetanHerve) [inline Modules/rating/Controller/RatingController.php:320]: Space managers or admins status can't evaluate ? So maybe we shouldn't send them campaign emails... And maybe allow campaign author to choose to include managers and / or admins

Comment (osallou) [reply Modules/rating/Controller/RatingController.php:320]: there is no sens for managers/admins to  evaluate. Satisfaction survey is for the users/customers. You don't give yourself a note....

but true, email is sent to users with a booking,... which could be a manager/admin. Their role should be checked for unnecessary emails

Comment (gaetanHerve) [reply Modules/rating/Controller/RatingController.php:320]: Sometimes, "users" can have manager status. But it results from a wrong usage of PFM. ok

Comment (osallou) [reply Modules/rating/Controller/RatingController.php:320]: Let's keep it like this for the moment, we will see on usage (if any...)

File: Modules/rating/Controller/RatingController.php [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+
+            $total_stats = $r->statGlobal($id_space, $id_campaign);
+            foreach ($total_stats as $stat) {
+                $stat['rate'] = round($stat['rate']);
+                $total[$stat['module']] = $stat;
+            }
+        }
+
+
+        return $this->render(['data' => ['answers' => $answers, 'total' => $total, 'global_projects' => $global_projects, 'global_bookings' => $global_bookings, 'campaign' => $data, 'bookings' => $booking_ratings, 'projects' => $projects_ratings], 'lang' => $lang, 'form' => $form->getHtml($lang)]);
```
Comment (gaetanHerve) [inline Modules/rating/Controller/RatingController.php:232]: $answers only defined i a else condition, line 129. Consequently not defined under certain circumstances.

File: Modules/rating/Controller/RatingController.php [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+                $this->request->setParam('from_date', $data['from_date']);
+                $this->request->setParam('to_date', $data['to_date']);
+                $this->request->setParam('message', $data['message']);
+                $_SESSION['flash'] = RatingTranslator::CampaignSaved($lang);
+
+            } else {
+                $_SESSION['flash'] = RatingTranslator::CampaignStarted($lang);
+            }
+
+            $from_date = strtotime($this->request->getParameter('from_date'));
```
Comment (gaetanHerve) [inline Modules/rating/Controller/RatingController.php:152]: returns false on updates.
this works: 
```
$from_date = strtotime($this->request->getParameter('from_date'));
            $to_date = strtotime($this->request->getParameter('to_date'));

            if($id_campaign) {
                $this->request->setParam('message', $data['message']);
                $_SESSION['flash'] = RatingTranslator::CampaignSaved($lang);
            } else {
                $_SESSION['flash'] = RatingTranslator::CampaignStarted($lang);
            }
```

File: Modules/rating/View/Rating/rateAction.php [hunk 4]
```diff
@@ -0,0 +1,76 @@
+<?php include 'Modules/core/View/spacelayout.php' ?>
+
```
Comment (gaetanHerve) [inline Modules/rating/View/Rating/rateAction.php:2]: still used ? Replaced by surveyAction.php, no ?

File: Modules/rating/View/Rating/surveyAction.php [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+                switch (kind) {
+                    case 'booking':
+                        this.resources[index].id = data.rate.id;
+                        break;
+                    case 'projects':
+                        this.projects[index].id = data.rate.id;
+                    default:
+                        break;
+                }
+            }).
```
Comment (gaetanHerve) [inline Modules/rating/View/Rating/surveyAction.php:123]: would be nice to add also a msg to confirm evaluation went ok.

**Review by gaetanHerve — COMMENTED** [review]
Seems ok to me appart from the comments I left.

File: Modules/rating/View/Rating/rateAction.php (hunk 4 above)
Comment (osallou) [reply Modules/rating/View/Rating/rateAction.php:2]: used by rateAction controller

File: Modules/rating/Controller/RatingController.php (hunk 3 above)
Comment (osallou) [reply Modules/rating/Controller/RatingController.php]: not what is expected but I pushed a fix

Comment (gaetanHerve) [reply Modules/rating/Controller/RatingController.php]: I saw that. tested and ok

**Pull request merged**